fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java

```diff
@@ -0,0 +1,58 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.collect.Lists;
+import com.starrocks.qe.SessionVariable;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+
+/**
+ * If there is only one BE and one fragment instance, we could generate one-phase local aggregation
+ * with the local shuffle operator (Scan->OnePhaseAgg->LocalShuffle) regardless of the differences
+ * between grouping keys and scan distribution keys.
+ */
+public class GenerateAggregateWithLocalShuffleRule extends TransformationRule {
+
+    protected GenerateAggregateWithLocalShuffleRule() {
+        super(RuleType.TF_GENERATE_AGGREGATE_WITH_LOCAL_SHUFFLE,
+                Pattern.create(OperatorType.LOGICAL_AGGR, OperatorType.PATTERN_LEAF));
+    }
+
+    private static final GenerateAggregateWithLocalShuffleRule instance = new GenerateAggregateWithLocalShuffleRule();
+
+    public static GenerateAggregateWithLocalShuffleRule getInstance() {
+        return instance;
+    }
+
+    public static boolean enable(SessionVariable sessionVariable) {
```

Comment:
Need we disable the two phase aggregate in the memo?

Reply:
Maybe not, when the aggregation cardinality is very low, two-phase is better than one-phase with local shuffle.